### PR TITLE
Documentation + Nix fixes

### PR DESCRIPTION
**Pull request by riz0id**
This PR contains several changes related to maintenance of Spectacle documentation, Haskell language server interoperability, and nix configuration changes. This PR builds off of the work in #52 which made nontrivial improvements to are documentation and Nix configuration. 

- Nix changes:
  1. Adds a `shell.nix` for Haskell language server compatibility 
  2. Factor out Nix overlays
- Haskell changes:
  1. Add `show-extensions` pragma to all Haskell modules
  2. Move internal Haskell modules to `other-modules` stanza
  3. Add haddock module descriptions to all Haskell modules
  4. Removed unused language pragmas
  5. Removed unused Haskell library modules and test modules
  6. Fix minor Haddock typos

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -13,29 +13,55 @@
...
+          };
+          haskell-language-server = { 
+            type = "app"; 
+            program = "${packages.haskell-language-server}/bin/haskell-language-server"; 
+          };
+        };
+
+        devShells = {
+          default = haskell-packages.shellFor {
+            name = "grpc-mqtt";
```
Comment (ixmatus) [inline flake.nix]:
```suggestion
            name = "spectacle";
```

File: flake.nix [hunk 2]
```diff
@@ -13,29 +13,55 @@
...
-        overlays = [ ];
-        pkgs = import nixpkgs { inherit system overlays; };
+        pkgs = import nixpkgs { 
+          inherit system; 
 
-        haskellPackages =
-          # Mangle the version string into the proper attribute name, then use that.
-          let attr = with pkgs.lib; "ghc" + (concatStrings (splitString "." ghcVersion));
-          in pkgs.haskell.packages."${attr}";
+          config.packageOverrides = import (nix/overlays.nix) {
```
Comment (ixmatus) [inline flake.nix:23]:
```suggestion
          overlays = import (nix/overlays.nix) {
```

File: flake.nix [hunk 3]
```diff
@@ -13,29 +13,55 @@
         system.aarch64-darwin
       ];
 
-      ghcVersion = "9.2.4";
+      ghc = "ghc924";
```
Comment (ixmatus) [inline flake.nix]:
```suggestion
      ghcVersion = "ghc924";
```

File: flake.nix [hunk 4]
```diff
@@ -13,29 +13,55 @@
...
+            type = "app"; 
+            program = "${packages.haskell-language-server}/bin/haskell-language-server"; 
+          };
+        };
+
+        devShells = {
+          default = haskell-packages.shellFor {
+            name = "grpc-mqtt";
+
+            buildInputs = with haskell-packages; [
```
Comment (ixmatus) [inline flake.nix:53]: Can we add `cabal-install` and `ghcid` for the weirdos like me that still aren't using HLS?

File: nix/overlays.nix [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# An attribute set containing arguments to be supplied to each overlay.
+args: 
+# The nixpkgs package set.
+pkgs:
+
+pkgs.lib.composeManyExtensions (map (f: f args) [
+  (import overlays/logict.nix)
```
Comment (ixmatus) [inline nix/overlays.nix]: Replace these two with `overlays/haskell.nix` to make it easier to maintain.

**Review by ixmatus — APPROVED** [review]

File: nix/overlays.nix (hunk 5 above)
Comment (riz0id) [reply nix/overlays.nix]: Haskell language server replaces `ghcid` and providing `cabal-install` in the shell can cause issues for GHC `9.4.2` and `9.2.4` in my experience.

File: flake.nix (hunk 2 above)
Comment (riz0id) [reply flake.nix:23]: I will open a separate PR for this fixing overlays and CI.

**Pull request merged**